clippy_lints/src/use_self.rs

```diff
@@ -54,235 +55,400 @@ declare_clippy_lint! {
+
+impl<'tcx> LateLintPass<'tcx> for UseSelf {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        // We push the self types of `impl`s on a stack here. Only the top type on the stack is
+        // relevant for linting, since this is the self type of the `impl` we're currently in. To
+        // avoid linting on nested items, we push `StackItem::NoCheck` on the stack to signal, that
+        // we're in an `impl` or nested item, that we don't want to lint
+        //
+        // NB: If you push something on the stack in this method, remember to also pop it in the
+        // `check_item_post` method.
+        match &item.kind {
+            ItemKind::Impl(Impl {
+                self_ty: hir_self_ty,
+                of_trait,
+                ..
+            }) => {
+                let should_check = if let TyKind::Path(QPath::Resolved(_, ref item_path)) = hir_self_ty.kind {
+                    let parameters = &item_path.segments.last().expect(SEGMENTS_MSG).args;
+                    parameters.as_ref().map_or(true, |params| {
+                        !params.parenthesized && !params.args.iter().any(|arg| matches!(arg, GenericArg::Lifetime(_)))
+                    })
+                } else {
+                    false
+                };
+                let impl_trait_ref_def_id = of_trait.as_ref().map(|_| cx.tcx.hir().local_def_id(item.hir_id));
+                if should_check {
+                    self.stack.push(StackItem::Check {
+                        hir_id: hir_self_ty.hir_id,
+                        impl_trait_ref_def_id,
+                    });
+                } else {
+                    self.stack.push(StackItem::NoCheck);
+                }
+            },
+            ItemKind::Static(..)
+            | ItemKind::Const(..)
+            | ItemKind::Fn(..)
+            | ItemKind::Enum(..)
+            | ItemKind::Struct(..)
+            | ItemKind::Union(..)
+            | ItemKind::Trait(..) => {
+                self.stack.push(StackItem::NoCheck);
+            },
+            _ => (),
+        }
+    }
+
+    fn check_item_post(&mut self, _: &LateContext<'_>, item: &Item<'_>) {
+        use ItemKind::{Const, Enum, Fn, Impl, Static, Struct, Trait, Union};
+        match item.kind {
+            Impl { .. } | Static(..) | Const(..) | Fn(..) | Enum(..) | Struct(..) | Union(..) | Trait(..) => {
+                self.stack.pop();
+            },
+            _ => (),
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &hir::ImplItem<'_>) {
+        // We want to skip types in trait `impl`s that aren't declared as `Self` in the trait
+        // declaration. The collection of those types is all this method implementation does.
+        if_chain! {
+            if let ImplItemKind::Fn(FnSig { decl, .. }, ..) = impl_item.kind;
+            if let Some(StackItem::Check { impl_trait_ref_def_id: Some(def_id), .. }) = self.stack.last().copied();
```

Comment:
Could you remove the `Option` from `impl_trait_ref_def_id` and just put `NoCheck` on the stack if it is `None`?

Also, I wonder if it would be better to avoid the stack altogether by using a separate Visitor which does not visit nested items. So, instead of `stack.push()`, you would have `MyItemVisitor().visit_item()` and other `check_` methods would be moved to `MyItemVisitor::visit_*`?

I wanted to bring up this idea because I also ran into a case where I need to have state-per-item and I was wondering which approach I should use. I hope that makes sense.

Reply:
> Could you remove the `Option` from `impl_trait_ref_def_id` and just put `NoCheck` on the stack if it is `None`?

Maybe? I'm not sure anymore, I implemented this ages ago. But I think you have to differentiate between trait impls and `NoCheck`s, because you have to prevent linting types in a trait impl, that are not defined as `Self` in the trait definition.

> Also, I wonder if it would be better to avoid the stack altogether by using a separate Visitor which does not visit nested items

The visitor method was used before. There are many other problems with that approach.  E.g. not being able to allow lints on expression level. Or that you still want to lint nested items and therefore still have to keep track of the latest state type -> stack. 

But the main problem I see with visitors is the redundant approach: You already are in a visitor context. Going into a visitor from there doesn't make much sense. (In general we should only use visitors to collect something, not to lint something)

Reply:
Thanks for explaining. I know you probably just want to get this PR out the door. 😃

Reply:
No problem. In fact, if you have more ideas or suggestions, please keep them coming. I'm petty much all out of ideas... :smile: